### DISCUSSION TITLE
EMQX as Alpine  Linux Service fails to stop

### DISCUSSION THREAD

**Discussion by cocoaway** [emqx/emqx#4622] [Q&A]
**Environment**:

- EMQ X version (e.g. `emqx_ctl status`):  4.3-rc.4-fe080333
- Hardware configuration (e.g. `lscpu`): Virtual machine 
- OS (e.g. `cat /etc/os-release`): Alpine Linux v3.13
- Kernel (e.g. `uname -a`): Linux alpv1 5.10.27-0-virt #1-Alpine SMP Wed, 31 Mar 2021 07:10:36 UTC x86_64 Linux
- Erlang/OTP version (in case you build emqx from source code): erlang-23.1.5-r0
- Others:

**What happened and what you expected to happen**:
OpenRC rc-service is able to start emqx application succesfully.
When requesting a stop service fail because emqx app has a different pid number from the one generated during start sequence.
Also ./emqx start takes up to nearly 30 seconds to start on some alpine linux virtual machines and more than 10 sec to stop.

**How to reproduce it (as minimally and precisely as possible)**:
Create a rc-service in a canonical way.

#!/sbin/openrc-run

description="emqx mqtt service !!!"

depend() {
  need net
}

start() {
  ebegin "Starting emqx broker : ${RC_SVCNAME}"
  start-stop-daemon --background --start   --exec /root/emqx/bin/emqx  start \
  --make-pidfile --pidfile /run/emqx.pid
  eend $?
}

stop() {
  start-stop-daemon --stop --exec /root/emqx/bin/emqx stop \
  --pidfile /run/emqx.pid
  eend $?
}

install in in etc/init.d with name emqx and make it executable.

upload it with 
`rc-update add emqx`

start it with
`rc-service start emqx`
>  * Starting emqx broker : emqx ...

emqx is running , now request for a stop
`rc-service emqx stop`
>* start-stop-daemon: no matching processes found

error is due to wrong pid created when service is starting.

In which way I have to write a rc-service that manages emqx with the pid created by emqx only after it started ?

**Anything else we need to know?**:

**Comment by zmstone** [comment]
Hi @cocoaway 
the `start` command will start emqx in daemon mode, the wrapping program is called `run_erl`.

instead of `start`, you can maybe try to start emqx in `foreground` mode?

Or try to `start-stop-daemon` the `run_erl` program.
not sure if it's the same for alpine, but here is a debian ref: https://github.com/emqx/emqx/blob/5c49d5e49ddf6d64554e8e073cc3be612b213abe/deploy/packages/deb/debian/init.script#L51-L74

**Reply by cocoaway** [reply]
Hi @zmstone , thanks for suggestions.
The issue with openrc service is the pid number.
The service start but openrc is unable to stop it and cause a crash.

Can you tell me the full command to be used with `run-erl `to start and stop emqx ?

Thank you.

**Reply by zmstone** [reply]
you can get the path as how `ERTS_PATH` variable is assigned in the example code.
in your setup, it should be

```
ERTS_PATH=`/root/emqx/bin/emqx ertspath`
```
then `run_erl` should have its full path at `$ERTS_PATH/bin/run_erl`

**Reply by cocoaway** [reply]
@zmstone  
I would like to use the shell script that comes with emqx located in /emqx/bin (emqx start - emqx stop)

But this way allows  to start emqx but not to stop it cleanly, in fact rc-service crashes.

I would like to ask for an rc-service script to be released to start and stop the emqx broker in Alpine, as already done for debian and redhat.

Can you provide a complete sample script for Alpine Linux?

Thank you.

**Reply by zmstone** [reply]
Hi @cocoaway 

I have added a feature request to track it.
https://github.com/emqx/emqx/issues/4640

Not sure when we can get the time for it though.

**Reply by cocoaway** [reply]
I could help you in testing the scripts you will make on Alpine Linux machines

**Comment by cocoaway** [comment]
Using Monit instead of rc-service solves the need, but it doesn't use openrc. 

In Alpine install it with : 
` apk add monit`
